### ISSUE TITLE
u-blox M10 GPS support 

### ISSUE THREAD

**Issue by Allerhopp** [mstrens/oXs_on_RP2040#18]
u-blox GPS with M10 chipset do not work anymore with oXs. Eventually this is because NAV-SOL messages are not supported.

From u-blox M10_ROM_5.00_ReleaseNotes_UBX-20050981:
5.2.3 Removed UBX messages

UBX-NAV-SOL PUB Use UBX-NAV-PVT instead

Is it possible to implement NAV-PVT instead or additionally?

**Comment by mstrens** [comment]
I tried to add UBX-NAV-PVT in a new version of the program.
Currently the change is only available in a new branch named "hott" (because it is supposed to support Hott protocol too).
I did not test the change (I have only a ublox M8).

Please test it and let me know if it is OK.
I will then put the version in master branch.

**Comment by Allerhopp** [comment]
GPS, GSpd, Hdg, GAlt and Tmp1 are working!

IIRC the M10 GPS was at 115200 Bd by default. I've used u-center 2 to set it to 9600 Bd. It would be good, if someone else could test a M10 GPS out of the box to verify the default baudrate.

**Comment by mstrens** [comment]
Thanks for the feedback.
I checked the M10 datasheet and it says that the default baudrate is 9600.
So I expect that oXs firmware should work for a M10 out of the box

**Comment by Allerhopp** [comment]
I've reopened this issue. It is still not sure from some posts in a German forum that 9600 works for out of the box M10 GPS.

**Comment by Satcomix** [comment]
Hallo Mstrens, 
The BeiTian Gps with M10 Chipset (BE-220) have a default baudrate of 38400 and must be set to 9600baud to work with oXS_on_RP2040
https://store.beitian.com/products/beitian-gps-module-with-antenna-ubx-m10050-gnss-chip-ultra-low-power-gnss-receiver-for-track-be-220
Greetings, 
Torsten

**Comment by mstrens** [comment]
I put a version 1.4.14 in test branch that tries to find the gps baudrate automatically (between 9600, 19200, 38400 and 115200).
This could probably solve this issue.

**Comment by doloebig** [comment]
i tested with B**E**-220 (new 38400Baud) is not detected (man. set to 9600 is working)
Blinks still with 1hz

KRegard 
Doro

**Comment by mstrens** [comment]
oXs is supposed to analyse the GPS frame during 600 msec (currently).
It could be , if the GPS is configured on 1Hz, that oXs does not get any frame during this interval.
I would perhaps have to increase the interval.

Can you make a test setting the GPS on 38400 bauds and 10Hz.
If oXs recognizes the baudrate, then the code is ok and just have to increase the interval.

**Comment by doloebig** [comment]
@mstrens 
for make it clear, if i set B**E**-220 manule to 9600 ->work, during startup will set to 2Hz 
If i leave default setting (38400) oXs will not register.

**Comment by mstrens** [comment]
can you try version 1.5.7 with gps being on default setting (38400)
I increased the delay to have more chance to detect the baudrate.
Please activate serial monitor at power up in order to see the first messages that will be generated.

You should get 
gps baudrate not detected; will be set on 9600
or 
gps baudrate detected = xxxxx) 
depending on the fact that baudrate is detected or not.

**Comment by doloebig** [comment]
works for me. I guess was timing issuew

Start detecting baudrate
gps baudrate detected = 38400
Attention: setup on core 1 did not ended within timeout
Setup1 takes 5000144 usec

Version = 1.5.7 
    Function                Pin   Change entering XXX=yyy (yyy=255 to disable)
Primary channels input    =  255  (PRI     = 5, 9, 21, 25)
Secondary channels input  =  255  (SEC     = 1, 13, 17, 29)
Telemetry . . . . . . . . =    7  (TLM     = 0, 1, 2, ..., 29)
GPS Rx  . . . . . . . . . =    3  (GPS_RX  = 0, 1, 2, ..., 29)
GPS Tx  . . . . . . . . . =    2  (GPS_TX  = 0, 1, 2, ..., 29)
Sbus OUT  . . . . . . . . =  255  (SBUS_OUT= 0, 1, 2, ..., 29)
RPM   . . . . . . . . . . =  255  (RPM     = 0, 1, 2, ..., 29)
SDA (I2C sensors) . . . . =  255  (SDA     = 2, 6, 10, 14, 18, 22, 26)
SCL (I2C sensors) . . . . =  255  (SCL     = 3, 7, 11, 15, 19, 23, 27)
PWM Channels 1, 2, 3 ,4   =  255  255  255  255 (C1 / C16= 0, 1, 2, ..., 15)
PWM Channels 5, 6, 7 ,8   =  255  255  255  255
PWM Channels 9,10,11,12   =  255  255  255  255
PWM Channels 13,14,15,16  =  255  255  255  255
Voltage 1, 2, 3, 4        =   26   27  255  255 (V1 / V4 = 26, 27, 28, 29)

Protocol is Sport (Frsky)
CRSF baudrate = 420000
Voltage parameters:
    Scales : 1.000000 , 25.862000 , 1.000000 , 1.000000 
    Offsets: 0.000000 , 13448.299805 , 0.000000 , 0.000000 
    No temperature sensors are connected on V3 and V4
RPM multiplier = 1.000000
Baro sensor is not detected
Acc/Gyro is not detected
First analog to digital sensor is not detected
Second analog to digital sensor is not detected
Foreseen GPS type is Ublox  :GPS is detected but has not (yet) a fix
Failsafe type is HOLD


THX
Doro

**Comment by mstrens** [comment]
Thanks for the feedback.
The current solution is still not optimal.
Testing the baudrate blocks the startup of oXs for some seconds.
This is not a real issue when oXs is used only for telemetry but when it is used also for generating Sbus and/or PWM signals it would be safier if oXs can be operational much faster.
I will probably try to find a solution to make the autobaudrate detection a non blocking process.

**Comment by mstrens** [comment]
Configuration of GPS U-blox M10 uses  different messages compared to M6/M8 and has a different default baudrate.
I made changes in version  1.9.0 in order to configure both types.
I tested it with a M8 but I do not have a M10 to test it.

**Comment by Satcomix** [comment]
Hello Mstrens,
I just installed the new version 1.9.0.
GPS M10 (Beitian BE-220/250 M10050) is not recognized.
With version 1.8.9 everything worked.
The new M10 GPS are also the reason why I haven't reported the last two weeks, because I've tried a lot of tests with different GPS (8030 and 10050) configs.
The config of an M10 is five times larger than that of the M8.
Best regards,
Torsten

**Comment by mstrens** [comment]
Perhaps that I made an error in the set of char I sent for configuring a M10.
Question: in u-center (or with another tool) are you able to send a set of char to the GPS and see if it replies with an ack or an nack.
If it reply with a nack I made a mistake in the set of char I am sending.
The set of char I send is:
0xB5,0x62,0x06,0x8A,   // config
        34, 0,  //length 4 + payload here after
        0x00,0x01,0x00,0x00,  // in ram
        0X01,0X00,0X21,0X30,   0X64 , 0X00, // key + Val in little endian for measurement rate (100 = 10Hz)
        0X2A,0X00,0X91,0X02,   0X01, // key + Val in little endian for POSLLH
        0X43,0X00,0X91,0X92,   0X01, // key + Val in little endian for VELNED
        0X07,0X00,0X91,0X92,   0X01, // key + Val in little endian for PVT
        0X02,0X00,0X74,0X10,   0X00, // L - - Flag to indicate if NMEA should be an output protocol on UART1
        0X3F,0X3E// checksum

**Comment by Satcomix** [comment]
Hello Mstrens,
I have attached some configurations of the BeiTian M10 BE-220/250
default config 38400 baud.
For the questions about ACK or NACK from you, I have to try something else.
[Beitian_BE_220_GPS_SBAS_Galileo_BeiDou_QZSS_config.txt](https://github.com/mstrens/oXs_on_RP2040/files/11069785/Beitian_BE_220_GPS_SBAS_Galileo_BeiDou_QZSS_config.txt)

[Beitian_BE_220_GPS_SBAS_Galileo_QZSS_GLONASS_config.txt](https://github.com/mstrens/oXs_on_RP2040/files/11069786/Beitian_BE_220_GPS_SBAS_Galileo_QZSS_GLONASS_config.txt)

[Beitian_BE_220_ONLY_UBX_in_out_NMEA_Child_messages_disabled_config.txt](https://github.com/mstrens/oXs_on_RP2040/files/11069787/Beitian_BE_220_ONLY_UBX_in_out_NMEA_Child_messages_disabled_config.txt)
br,
Torsten

**Comment by Satcomix** [comment]
Hello Mstrens,
I copied the config values ​​from the U-Center Config-Terminal for the specific settings. I hope it helps.

[M10_Config_Values.txt](https://github.com/mstrens/oXs_on_RP2040/files/11069862/M10_Config_Values.txt)

br,
Torsten

**Comment by mstrens** [comment]
Can you try version 1.9.3.
I found some mistake in the code being sent.
Here the sequence I sent now (to use if you find a way to send a sequence and listen to ack/nack:
    0xB5,0x62,0x06,0x8A,   // config
        34, 0,  //length 4 + payload here after
        0x00,0x01,0x00,0x00,  // in ram
        0X01,0X00,0X21,0X30,   0X64 , 0X00, // key + Val in little endian for measurement rate (100 = 10Hz)
        0X2A,0X00,0X91,0X20,   0X01, // key + Val in little endian for POSLLH
        0X43,0X00,0X91,0X20,   0X01, // key + Val in little endian for VELNED
        0X07,0X00,0X91,0X20,   0X01, // key + Val in little endian for PVT
        0X02,0X00,0X74,0X10,   0X00, // L - - Flag to indicate if NMEA should be an output protocol on UART1
        0X79,0XC6// checksum

**Comment by Satcomix** [comment]
Here are some more informations about GNSS config.
UBX-CFG-GNSS GPS 8/16 L1C/A, SBAS 3/3 L1C/A, Galileo 8/12 E1, QZSS 3/4 L1C/A L1S, GLONASS 8/12 L1OF, Number of channels: 49
0000  B5 62 06 3E 34 00 00 00 31 06 00 08 10 00 01 00 01 01  µb>41
0012  01 03 03 00 01 00 01 01 02 08 0C 00 01 00 01 01 03 06  
0024  0E 00 00 00 00 01 05 03 04 00 01 00 05 01 06 08 0C 00  
0036  01 00 01 01 35 58                                      5X

I will make the next test with 1.9.3.
br,
Torsten

**Comment by Satcomix** [comment]
Hello Mstrens,
Sorry,Version 1.9.3 **no GPS** in terminal, FV or handheld
br,
Torsten

**Comment by mstrens** [comment]
FIY, I saw in the M10 datasheet that old configuration messages (UBX_CFG...) are deprecated
"This message is deprecated in protocol versions greater than 23.01. Use UBX-CFG-VALSET, UBX-CFGVALGET, UBX-CFG-VALDEL instead."

That is why I try to use UBC_CFG_VALSET and I sent the command at 38400 baud (which is the default rate for M10).
There must be a mistake somewhere but as I have no M10 gps I can't test it.

Perhaps can you make a capture with a logic analyser of the TX and the Rx GPS signals when GPS and OXS are powered on.

**Comment by Satcomix** [comment]
Hello Mstrens,
the only way to see ack or nack would be with a logic analyzer, right?
I could also connect the RX and TX lines from the oXs_RP2040 to the GPS in parallel to the U-Center, I just don't know if I can see these confirmation values ​​(ack/nack).
I try the way with RX/TX parallel to the U-Center first and than i will make a capture with the Kingst.
br,
Torsten

**Comment by Satcomix** [comment]
Kingst capture:
one TX, one RX capture
[2023-0
[2023-03-25_21-25-01.zip](https://github.com/mstrens/oXs_on_RP2040/files/11070006/2023-03-25_21-25-01.zip)
3-25_21-24-19.zip](https://github.com/mstrens/oXs_on_RP2040/files/11070005/2023-03-25_21-24-19.zip)

**Comment by mstrens** [comment]
I presume that the timelaps of your capture is to short.
I do not see the messages sent by oXs.
Best is to capture on 10 sec about.

**Comment by Satcomix** [comment]
The problem with the Kingst Capture is that the oXs Tx line switches off and no more data is visible. The RX line receives the data from the GPS all the time.
But we already had this problem with the GPS measurements in the last few weeks.

**Comment by mstrens** [comment]
Sorry, I did not saw the second capture. In the second one the capture seems to start to late.
In fact both signals should be in the same capture (if possible) and for 10 sec from power on.
Is this possible

**Comment by Satcomix** [comment]
Next try: i hope i stopped at the right moment.
[2023-03-25_21-40-24.zip](https://github.com/mstrens/oXs_on_RP2040/files/11070040/2023-03-25_21-40-24.zip)

**Comment by mstrens** [comment]
If Kingst can capture 2 channels simultenously I could
- first find at the first messages sent by oXs
- see if just after the first message, the GPS reply (and so see if it is an Ack or a NACK reply

**Comment by Satcomix** [comment]
ok, i will make two new captures one RX, one TX.

**Comment by mstrens** [comment]
I saw both signals on the last capture. It should be OK.
Still the capture start to late.
Can you e.g. trigger the capture base on the signal sent by the receiver. This one exist before the signal sent by oXs.
Or (easier) start the capture manually just before powering on OXS/GPS

**Comment by Satcomix** [comment]
I always start the capture before i power on oXs_RP2040
[2023-03-25_21-52-48.zip](https://github.com/mstrens/oXs_on_RP2040/files/11070113/2023-03-25_21-52-48.zip)
[2023-03-25_21-50-18.zip](https://github.com/mstrens/oXs_on_RP2040/files/11070114/2023-03-25_21-50-18.zip)

**Comment by mstrens** [comment]
I do not understand what happens
The first capture is "empty" (no signal)
The second contains messages sent by oXs but not the first part of messages.

I will make a debug version that will send to the PC a copy of what is sent to the GPS.

**Comment by Satcomix** [comment]
It was a long day and i will make the next Tests tomorow.
br,
Torsten

**Comment by mstrens** [comment]
OK.
Thanks for testing.
FYI, I made a capture myself and I see that oXs sent the expected frame.
Still as I am using a M8 with 9600 default baud, I do not see the ACK/ACK when I sent the first sentence (for M10) at 38400 bauds.

**Comment by Satcomix** [comment]
Hello Mstrens,
The M10 GPS is not initialized with version 1.9.7, and drives with default 38400baud and 1Hz, without connection to the oXs (Foreseen GPS type is Ublox :GPS is not (yet) detected)
Wouldn't it be possible to read the configuration commands in HEX with a connected M10 GPS in the U-Center at Messages View (not Configuration View) and enter them into oXs Config?
Greeting,
Torsten

**Comment by mstrens** [comment]
With an M8 (or previous) I was able to send configuration commands to the GPS with U-Center and to see the hex message being sent by U-center. That is the way I proceeded previously.
This works only if a GPS is connected to U-Center.
As I do not have a M10, I can't try it with a M10.
I will look at the library you found. Perhaps I will see the issue.

Yesterday I connected my logic analyser to oXs and I started a capture for 10 sec.
I clearly saw that oXs sent the expected messages (first a short messages at 38400 and then several messages at 9600).
This seems OK but I do not know if the content of the first message (for M10) was valid or not.

**Comment by Satcomix** [comment]
Hello Mstrens,
I have connected several M10s here.(BE-180/220/250/280)
If you want, I could send you one to your home for free, it only takes 3 days from GE to BE. Just a suggestion.
I could also read out the hex messages that are to be sent in the Message Viewer, as I did yesterday, but then I would only have to know which ones exactly.
Greeting,
Torsten

**Comment by mstrens** [comment]
If you can read out the hex message, then it should be possible to move on.
Currently I assume that the GPS is in default config and I send 5 commands
All commands are supposed to be loaded in the RAM layer.
The 5 commands are (name , key, value , comment):
CFG_RATE_MEAS 0x30210001 100  // set measurement rate to 10Hz
CFG_MSGOUT_UBX_NAV_POSLLH_UART1 0x2091002a  1 // ask to output a UBX-NAV-POSLLH message on port UART1 (at 10hz)
CFG_MSGOUT_UBX_NAV_VELNED_UART1 0x20910043  1 // ask to output a UBX-NAV-VELNED message on port UART1 (at 10 hz)
CFG_MSGOUT_UBX_NAV_PVT_UART1    0x20910007 1 // ask to output a UBX-NAV-PVT message on port UART1 (at 10 Hz)
CFG_UART1OUTPROT_NMEA 0x10740002 0 // ask to stop sending NMEA msgon UART1

        0x00,0x01,0x00,0x00,  // in ram
        0X01,0X00,0X21,0X30,   0X64 , 0X00, // key + Val in little endian for measurement rate (100 = 10Hz)
        0X2A,0X00,0X91,0X20,   0X01, // key + Val in little endian for POSLLH
        0X43,0X00,0X91,0X20,   0X01, // key + Val in little endian for VELNED
        0X07,0X00,0X91,0X20,   0X01, // key + Val in little endian for PVT
        0X02,0X00,0X74,0X10,   0X00, // L - - Flag to indicate if NMEA should be an output protocol on UART1

**Comment by Satcomix** [comment]
Hello Mstrens,
I read the values ​​in the U-Center at VALGET. I hope it fits.
Or should I read out the values ​​at VALSET?
No problem, I'm on it.
br,
Torsten
[M10_VALGET_RAM.txt](https://github.com/mstrens/oXs_on_RP2040/files/11071185/M10_VALGET_RAM.txt)

**Comment by mstrens** [comment]
The idea is to get the hex values for valset commands to compare them with what I sent.

For a full test, you could also:
- reset the GPS to default config
- send the command to avoid all output NEMA msg on UART1 (so U-center should then be quiet)
- send the command to activate one of the NAV msg (e.g. CFG_MSGOUT_UBX_NAV_POSLLH_UART1 0x2091002a 1 )
You should then get only this type of message in U-center.
If this works, you could send the other commands to change the rate and get other NAV msg.

**Comment by Satcomix** [comment]
I reset the M10 to DEFAULT values.
After that I ran NMEA disable Child messages. (No NMEA output to U-Center)
Where do the values ​​have to be stored with VALSET?
Set values ​​in layers:
FLASH
BBR
RAM
default is RAM
When I enter the 10Hz(100=64hex) I no longer see a blue LED, so something is wrong.
Key Name	Key ID	Type	Value
CFG-RATE-MEAS 	0x30210001	U2	100  0x64  s  scaled 0.001
0000  B5 62 06 8A 0A 00 01 01 01 00 01 00 21 30 64 00 53 CB     µb!0dSË
Than SEND

**Comment by Satcomix** [comment]
I found a problem.
If I disable all NMEA child messages and enable UBX messages, U-Blox only shows UBX, right!
But when I execute the first command (Config-Rate_Meas 10Hz) NMEA starts again!
How can I switch off the NMEA messages permanently stored?

**Comment by mstrens** [comment]
Perhaps you have to execute Config-Rate_Meas 10Hz first followed by the others commands.

Looking at your HEX trace, I found a difference in the way the payload was calculated.
I made a new version 1.9.8. (on github)
Perhaps this will solve the issue.

**Comment by Satcomix** [comment]
You got it!!!
It works with Ver.1.9.8 with 38400 and 10Hz
**Foreseen GPS type is Ublox  :GPS is detected and has a fix**
In Handheld everything seems ok.
Thank you for your work and time.
br,
Torsten

**Comment by mstrens** [comment]
Thanks too.
It is your trace that helped to find the bug.

**Comment by Satcomix** [comment]
Hello Mstrens,
I've looked at the M10's UBX output and it seems to me that the M10's default satellites are transmitting the messages.
I receive BeiDou instead of GLONASS.
UBX-CFG-GNSS 
GPS 8/16 L1C/A, 
SBAS 3/3 L1C/A, 
Galileo 8/12 E1, 
QZSS 3/4 L1C/A L1S, 
GLONASS 8/12 L1OF, 
Number of channels: 49

With BeiDou and without GLONASS (include GPS,SBAS,Galileo,QZSS)
0000  B5 62 06 3E 34 00 00 00 31 06 00 08 10 00 01 00 01 01 01  µb>41
0013  03 03 00 01 00 01 01 02 08 0C 00 01 00 01 01 03 06 0E 00  
0026  01 00 01 01 05 03 04 00 01 00 05 01 06 08 0C 00 00 00 00  
0039  01 35 78                                                  5x

With GLONASS and without Beidou (include GPS,SBAS,Galileo,QZSS)
0000  B5 62 06 3E 34 00 00 00 31 06 00 08 10 00 01 00 01 01 01  µb>41
0013  03 03 00 01 00 01 01 02 08 0C 00 01 00 01 01 03 06 0E 00  
0026  00 00 00 01 05 03 04 00 01 00 05 01 06 08 0C 00 01 00 01  
0039  01 35 58                                                  5X

br,
Torsten

**Comment by mstrens** [comment]
You can find the default config at the end of this doc
https://content.u-blox.com/sites/default/files/u-blox%20M10-SPG-5.00_InterfaceDescription_UBX-20053845.pdf

**Comment by Satcomix** [comment]
OK,
this is apparently the difference to the original U-Blox.
With the Beitian (M10) BE chipset it looks like this:
DEFAULT Values:
Configuration item Key ID Type Scale Unit Default value
CFG-SIGNAL-GPS_ENA 0x1031001f L - - 1 (true)
CFG-SIGNAL-GPS_L1CA_ENA 0x10310001 L - - 1 (true)
CFG-SIGNAL-SBAS_ENA 0x10310020 L - - 1 (true)
CFG-SIGNAL-SBAS_L1CA_ENA 0x10310005 L - - 1 (true)
CFG-SIGNAL-GAL_ENA 0x10310021 L - - 1 (true)
CFG-SIGNAL-GAL_E1_ENA 0x10310007 L - - 1 (true)
CFG-SIGNAL-BDS_ENA 0x10310022 L - - 1 (true)
CFG-SIGNAL-BDS_B1_ENA 0x1031000d L - - 1 (true)
CFG-SIGNAL-QZSS_ENA 0x10310024 L - - 1 (true)
CFG-SIGNAL-QZSS_L1CA_ENA 0x10310012 L - - 1 (true)
CFG-SIGNAL-GLO_ENA 0x10310025 L - - 0 (false)
CFG-SIGNAL-GLO_L1_ENA 0x10310018 L - - 1 (true)

oXs-config values:
Configuration item Key ID Type Scale Unit Default value
CFG-SIGNAL-GPS_ENA 0x1031001f L - - 1 (true)
CFG-SIGNAL-GPS_L1CA_ENA 0x10310001 L - - 1 (true)
CFG-SIGNAL-SBAS_ENA 0x10310020 L - - 1 (true)
CFG-SIGNAL-SBAS_L1CA_ENA 0x10310005 L - - 1 (true)
CFG-SIGNAL-GAL_ENA 0x10310021 L - - 1 (true)
CFG-SIGNAL-GAL_E1_ENA 0x10310007 L - - 1 (true)
CFG-SIGNAL-BDS_ENA 0x10310022 L - - 0 (false)
CFG-SIGNAL-BDS_B1_ENA 0x1031000d L - - 1 (true)
CFG-SIGNAL-QZSS_ENA 0x10310024 L - - 1 (true)
CFG-SIGNAL-QZSS_L1CA_ENA 0x10310012 L - - 1 (true)
CFG-SIGNAL-GLO_ENA 0x10310025 L - - 1 (true)
CFG-SIGNAL-GLO_L1_ENA 0x10310018 L - - 1 (true)
br,
Torsten

**Comment by Satcomix** [comment]
Hello Mstrens,
I put Version 1.9.9 on my board with a M8030 GPS , but it doesn't initialize properly and blinks in 1Hz.
Should I open a new issue for M8 GPS or shall we continue here?
Greetings,
Torsten

**Comment by mstrens** [comment]
best is to keep this because issue with M8 and M10 are related.
Trying to solve an issue with M8, I (re) created an issue with M10.
Probably I have to sent the M10 sentence at the end (after M8 sentence).

**Comment by Satcomix** [comment]
Ok, than we stay here.
The M10  initialize, and everything is shown in FV and Handheld with S.PORT/FBUS Protocol.

**Comment by mstrens** [comment]
I put version 1.9.10 on github.
I sent the M10 initialization sentence at the end.
Perhaps it will help that initialization is OK

**Comment by Satcomix** [comment]
Hello Mstrens,
Excuse the late answer.
Version 1.9.10 behaves the same as version.1.9.9. No improvement.

**Comment by mstrens** [comment]
I do not really understand.
I found 2 errors in checksum of 2 commands and I increase some delay between comands
I made a new version 1.9.11
It works with my M8
With the log analyser I see clearly each ACK/NACK after each command.
Hoping it works with M10 too.
If not, can you make one capture with both TX and RX signals to see if the Ack/nack are present.

**Comment by Satcomix** [comment]
Hello Mstrens,
With the M10 M10050) all values ​​are correctly transmitted.
With the M8 (M8030KT) only after a RESET.

**Comment by mstrens** [comment]
Very strange because with my M8, it works always after a power down.
Can you make a capture with M8 and power down from both TX and RX signals for at least 10 sec.

**Comment by Satcomix** [comment]
Hello Mstrens,
It's very peculiar.
I could observe this behavior:
After some time, oXs TX switches to high level and then the small initialization packets come. It worked on one out of three attempts, and the M8 initialized normally after power down.
The capture shows the normal ini of M8 after power down.
The next capture, i will try to get TX and RX from the failure.
[2023-03-27_18-03-37.zip](https://github.com/mstrens/oXs_on_RP2040/files/11080766/2023-03-27_18-03-37.zip)

**Comment by Satcomix** [comment]
Next try.
[2023-03-27_18-14-46.zip](https://github.com/mstrens/oXs_on_RP2040/files/11080803/2023-03-27_18-14-46.zip)

**Comment by Satcomix** [comment]
When I look at the capture under normal operation (38400baud) I see framing error. If I then switch to Autoframe, I get 38284 baud, is that where the problem lies?

**Comment by mstrens** [comment]
I do not think that the issue is related to the small baudrate difference.
Normally  a few % is not an issue to decode uart.

**Comment by mstrens** [comment]
Your captures does not cover a minimum time lap of 10 sec after power on or reset.
So It is not possible to see the first frame sent by oXs and the reply (if any) of the GPS.

**Comment by mstrens** [comment]
If the ini is accepted at some stage, it seems me normal that the GPS still continue to work after a reset of RP2040 because: 
- or the following commands (after reset) are accepted and so it works
-  or the following commands are rejected/discarded ans do it continue to works.

**Comment by Satcomix** [comment]
This is the problem of the Kingst analyzer or the oXs board (or me, because I set something wrong 20MSamp/100MHz), I can only always record the frame that is just recognized, e.g. Line2 GPS TX. If the Line1 is set to 0/ZeroVolt I can no longer save the frames with GPS TX. And oXs TX only sends a few frames while the GPS sends frames on 38400 all the time.

**Comment by Satcomix** [comment]
**You wrote:** or the following commands are rejected/discarded ans do it continue to works.
Then the M8 would have to continue running at 9600 baud, but it initializes to 38400 after a reset.

**Comment by mstrens** [comment]
perhaps you can try version 1.9.12.
I do not send anymore a change of GNSS init.
I read that from version 3 firmware, this generates a reset of the GPS.

**Comment by mstrens** [comment]
In version 1.9.12  with a M8 GPS,  does it means that the GPS stays always on 9600 even after a reset of the RP2040?

If so, I have perhaps to force a "reset" of the GPS at the end of the initialisation (in v3.01 firmware it seems that it is what happens with a GNSS config command.

**Comment by Satcomix** [comment]
My GPS types are 
TBS8-2 SPG2.01, 32SV
BN-220 SPG 3.1 Prot15(no Galileo), 32SV
BN-220 SPG3.1 Prot18 (Galileo), 32SV
BE-180/220/250 SPG5.10 Prot34.10 49SV
BE-280 SPG5.0 /Prot 3xxx 49SV
So I could test all SPG and PROT versions. 63SV
br,
Torsten

**Comment by mstrens** [comment]
now version 1.9.13.
I read that in some gps version, the Rx line is blocked when to many errors occurs.
It takes 1 sec to reset the Rx.
I now inserted a delay of 2 sec when I change from one baudrate to another.
Perhaps it help.

**Comment by Satcomix** [comment]
Now there is no communication for about 2 seconds at the beginning (blue LED off), after that no init, after a RESET an initialization takes place again, with 38400baud/10Hz

**Comment by mstrens** [comment]
Here a version 1.9.14 that can work only for M8 (to test step by step, I removed M10 init sequence).
Here the expected process:
At power on, GPS sent nema frame at 9600 and so led is blinking. 
After a few sec, oXs sent a command at 9600 baud that cancels the nema Msg and asks for switching to 38400 sec.
I imagine that GPS accept this command and led stops blinking for 2 sec (no nema msg anymore).
After 2 sec, oXs sent at 38400 baud a set of command to request 10 hz and some UBX messages for nav. So led should start blinking again. Frames sent by GPS should be at 38400 and 10HZ.

**Comment by Satcomix** [comment]
No change.
After power down the blue LED flashes 1Hz, after RESET the same way.
You could always see the beginning of the init, since the blue LED is still for this period, it now also does , but again does not receive an ack
Terminal: Foreseen GPS type is Ublox  :GPS is not (yet) detected

**Comment by Satcomix** [comment]
Next Test tomorow :-)
br,
Torsten

**Comment by mstrens** [comment]
I do not understand.
With my M8 it works as expected.


Ok for next tests tomorrow.
If the led is still for 2 sec, it means (I expect) that the GPS accepts the command asking for stopping the NEMA msg.
It is the same command that asks for switching to 38400. So why should this part of the command not be accepted????
On my capture, I clearly see that RX lines is floating at power on, it becomes high when oXs waited several sec after power on and begin to drive the pin,  then RX remains high for 2 sec and then oXs sent a first command at 9600 baud and get a ACK.
Perhaps I have to send commands to save the config on BRR and to ask for a soft reset?????

**Comment by pierrotm777** [comment]
It's my analyse with the 1.9.14 firmware

[gps_init3.zip](https://github.com/mstrens/oXs_on_RP2040/files/11086546/gps_init3.zip)

**Comment by Satcomix** [comment]
Hello Mstrens,
Hello Pierre, nice that you also test and help
I continued testing this morning.
Version 1.8.7:
The GPS M8 will initialize as requested.
The M10 also initializes with 38400 baud and 10Hz, with the red PPS LED only flashing very late, after a 3D fix with 9-10 SV.
Version 1.9.8:
The M10 initializes as requested.
The M8 does not initialize itself after power down, but only after a RESET.
Version 1.9.14:
The M8 behaves like version 1.9.8
After power down no ini, after reset there is an ini, but not always.
The M10 initialized with 1.9.8 only delivers NMEA data at first and changes to NMEA and UBX data after the ini at 38400baud/10Hz.
br,
Torsten

**Comment by pierrotm777** [comment]
Good news,
With 1.9.14 and after a reset, i can see my M8N :-)

**Comment by Satcomix** [comment]
> Good news, With 1.9.14 and after a reset, i can see my M8N :-)

Hello Pierre,
I've noticed that too, see the post above, but the reset doesn't always work at M8, and that's where the problem lies, otherwise Michel could install a reset at the beginning and wait for the ack.
Greetings,
Torsten

**Comment by mstrens** [comment]
@pierre,
in your capture, it is clear that the GPS does not recognize the first command that oXs sent at 9600baud. There is no ack/nack.
That is the issue.

Can you try the version 1.9.15. I increased the delay before sending the first command and I sent one another command before asking for the change of baudrate. 
Please make a capture taking care of:
- first upload
- make power off of RP2040 and GPS
- - make power on of both and start the capture for at least 15 sec.
-

**Comment by Satcomix** [comment]
It works with M8 and Ver.1.9.15.
Very long blue led at startup after power down and then init with 38400/10Hz

**Comment by mstrens** [comment]
I made 2 changes (lincrease a delay from 2 up to 5 sec, send another command before changing the baudrate).
I do not know what is really required.
The capture of Pierre will perhaps says it.
Perhaps there is a glitch in the signal that is so short that the logic analyser do not see it but that GPS can detect?
In a future version, I will also sent a dummy character and wait a little before sending the first command. 

On my side, I noticed that the logic analyser some time detect a glitch (and so a UART error) on the first byte sent by the GPS after a baudrate change. This issue is in the GPS (because oXs just listen on this pin). Anyway this does not have an impact on the fact that some commands sent by oXs are not recognized.

**Comment by mstrens** [comment]
In 1.9.15, 
- first GPS start sending msg (it starts before oXs drive the GPS RX line)
- after some delay (a few sec), oXs drive the GPS RX line HIGH for 5 sec (instead of 2)
- then oXs sent 2 commands; the first ask to send one UBX msg; the second ask to stop NEMA, start UBX Msg and change baudrate to 38400. (In previous version there was no command to ask sending a UBX msg)
- if those command are accepted, oXs start sending one (short) UBX msg at still 1Hz (so led should blink short and slowly)
- 2 sec after, oXs starts sending other commands ( asking for 4 types of UBX msg and for 10Hz) and so led should blink must faster

**Comment by pierrotm777** [comment]
With 1.9.15,
- 1st test is normal, no reset.
- 2sd test with start normal and reset after the normal start.
[startnormar&reset.zip](https://github.com/mstrens/oXs_on_RP2040/files/11087476/startnormar.reset.zip)

**Comment by pierrotm777** [comment]
Curious thing, if i add a decoder feature on the Tx AND Rx i see 9600 for one and 38400 for the other !
I use AutoBaud mode.
[rxtx_with_diff_baud.zip](https://github.com/mstrens/oXs_on_RP2040/files/11087547/rxtx_with_diff_baud.zip)

**Comment by pierrotm777** [comment]
As Satomix said, it's ok for me too with the 1.9.15.
Only after the message that follow, i can see my gps.
 check sum for   6  0  14  0  1  0  0  0  d0  8  0  0  0  96  0  0  7  0  1  0  0  0  0  0 
 is 91  84
lenght to specify is 20

**Comment by Satcomix** [comment]
The ACK response should be this:  B5 62 05 01 02 00 06 01 0F 38 µb.......8.
I've just made a capture several times, by being able to see who is sending and receiving something, how, and when. Unfortunately, GPS RX cannot record, is discarded by GPS TX after each new frame.
THE ACK from the GPS is sent up to eight times, after which it switches to UBX

**Comment by mstrens** [comment]
It seems that:
- at power on, OXS does not drive the GPS RX line. This line is floating and get a lot of false signal.
- in GPS firmware 3.xx there is a new feature that blocks the Rx line after 100 frames errors.
- I expect that the Rx line becomes active again after a quite long delay. I tested driving the line high for 2 sec before sending the first commands but it seems it was not enough.
- Increasing the delay up to 5 sec seems to solve the issue.

As I made several changes, I will now make a version with only the delay increase to confirm my expectation.

**Comment by mstrens** [comment]
I put version 1.9.17 
It keep RX line high 5 sec before sending first command.
First command disables NEMA , activates UBX and set baud on 38400.
oXs wait for 2 sec (perhaps not really needed - could be deleted in next version); so normally led shoudl not blink for 2 sec
oXs sent commands (asking 4 UBX msg, setting on 10 hz)

**Comment by Satcomix** [comment]
With version 1.9.17 the ini works without any problems.
It behaves from the LED exactly as you described it.

**Comment by pierrotm777** [comment]
Ok for me too with the 1.9.17 :-)

**Comment by mstrens** [comment]
Now version 1.9.18 where I reduce the delay of 2 sec to 4 msec (between first command and next commands.
If ok, I will then add code for M10 configuration.

**Comment by Satcomix** [comment]
Version 1.9.18 works with M8 with the reduce of the delay to 4 msec.

**Comment by pierrotm777** [comment]
It's ok for me too :-)

**Comment by pierrotm777** [comment]
It's ok with my M8N !

**Comment by Satcomix** [comment]
Version 1.9.20:
M8  works 38400baud/10Hz  **GPS and GLONASS**
M10 works 38400baud/10Hz **GPS , Galileo and BeiDou**
So far it works, the question is do you want to implement NMEA Ver 4.10 so that Galileo can also be received on the M8 SPG 3.01 PROTVERSION 18 and on the M10 SPG 5.10 PROTVERSION 34.10 i prefer the GLONASS SVs instead of BeiDou.
br,
Torsten

**Comment by pierrotm777** [comment]
M8 not accept Galileo ?

**Comment by Satcomix** [comment]
@Pierre
BN-220 M8030KT SPG 3.01 PROT 18 can receive Galileo SV with a specific setting (U-Center CFG NMEA4.10). 
But I think I had already explained it to you in the RCG.
Greeting,
Torsten

**Comment by pierrotm777** [comment]
If i can have galileo, ok for nmea

**Comment by Satcomix** [comment]
Adding Galileo Constellation
https://oscarliang.com/gps-settings-u-center/
The M8 chip is capable of receiving signal from several constellations of satellites, including:

    GPS (USA) – 31 Satellites
    GLONASS (Russia) – 23 satellites
    GALILEO (Europe) – 24 satellites
    BEIDOU (China) – 44 satellites

You can select up to 3 systems, and usually Galileo isn’t enabled by default, so we want to use that to increase the number of satellite we can see.

**Comment by mstrens** [comment]
In the datasheet for M8, it seems that SBAS and QZSS are also enabled by default. So best is to disable them I expect.
I read somewhere (ublox doc) that when you activate Galileo you should reduce the refresh rate (best max 3Hz?).
The provided links does not seems to say that refresh rate should be reduced.
If Galileo is activated, will it be integrated into the UBX messages being already used? I would like to avoid having to decode NEMA msg.

**Comment by Satcomix** [comment]
> In the datasheet for M8, it seems that SBAS and QZSS are also enabled by default. So best is to disable them I expect. I read somewhere (ublox doc) that when you activate Galileo you should reduce the refresh rate (best max 3Hz?). The provided links does not seems to say that refresh rate should be reduced. If Galileo is activated, will it be integrated into the UBX messages being already used? I would like to avoid having to decode NEMA msg.

The refresh rate can(must) be reduced by activate SBAS with 4 or 3 channels to 4Hz for accuracy.
I will make some tests with higher rates(115200baud) and SBAS 3Channel and 5Hz/10Hz.
As far as I understood you have to set in U-Blox Config the NMEA4.0 to NMEA4.10 in order to enable Galileo on FW SPG 3.01PROT18, also when you work with UBX messages.
In Fw SPG 5.10 (BE-220/250) the Galileo is default in the UBX message. I normally use the BE-250, it very accurate, but a liile bit bigger then BE-220.
Normally you activate GPS,SBAS,Galileo und GLONASS for Europe.
br, 
Torsten

**Comment by Satcomix** [comment]
In my opinion, nothing in the baud rate would have to be changed, since only a few messages are transmitted.
The 5Hz option could be offered for the nav rate, with simultaneous reception of 4 GNSS systems.
In addition, not everything has to be specified, you can make your own settings and save them in the GPS
In the Message Viewer (U-Center) you can see the utilization of the port under UBX MON-COMMS/TX-Buffer when NMEA messages are deactivated and some UBX messages are activated.
Would increasing the baud rate, e.g. from 38400baud to 57600 or 115200baud, put too much strain on the oXs_RP2040 system?
br,
Torsten

**Comment by Satcomix** [comment]
Thank you very much for your explanations.
I'm doing a few more tests with Galileo and SPG3.01PROT18.
NMEA child messages disabled, then UBX messages enabled.
With BN-220 I don't see Galileo, with BE-220 directly.
I'm getting more and more enthusiastic about the BE series(faster in processing the data and more accurate).
I'll keep trying, with or without NMEA4.10 :-)
If there's nothing else to do for me with the M8 or M10, I'll continue testing now.
Thank you for your work and effort.
br,
Torsten

**Comment by fourfingers4** [comment]
Hello all, and thanks for all the work.
Would like to help, but don't have access to my hardware until next week.
I understand some ublox receivers are unable to permanently store the configuration, so configuration from oXs is the best option.
 But many others can store the configuration, so I would really like oXs to also offer an option to set up ublox receivers the same way it does for Casic gps. 
Since ublox has so many different generations, with so many different options depending on firmware, prot version, etc., it would be much easier for the user to comply to certain requierements from oXs (protocol messages, baudrate...) and set their units to comply given whatever his gnss hardware possibilities/limitations. This will give the option to users to set up their hardware in the best possible way. Having oXs in charge of the task of aquainting for every possible ublox hw/fw combination without conflict seems quite complicated, at least without trade-offs. 
I really believe this will definitely  make things much easier, 
faster and better, at least in some cases, but just my point of view.

**Comment by Satcomix** [comment]
> This could be done. In order to avoid a new command to be used/displayed/stored from the PC serial terminal, I could add another GPS type. Currently there is 2 values: U (ublox) and C(cadis) I could add e.g. the value E (for Ublox Externally configured).

This is a good and practical idea. 
As I have already written, everyone can configure their own GPS according to their own interests.
br,
Torsten

**Comment by Satcomix** [comment]
Version 2.0.0 works with GPS M10 with E setting in terminal.
Foreseen GPS type is Ublox **(configured by oXs**) :GPS is detected and has a fix
So an **E** means that the GPS happens through the settings in the oXs Config.
A **U** means that the settings in the GPS are made by e.g. U-Center. 
Or have I misunderstood something?
br,
Torsten

**Comment by mstrens** [comment]
When you select the "E" option for GPS, oXs does not try to initialize the Ublox GPS.
If the GPS is sending data at 38400 baud (= default for a M10) and if at least following messages are sent it should be OK.
- UBX-NAV-POSLLH
- UBX-NAV-VELNED
- UBX-NAV-PVT
Best is also to disable all other messages(they are just not handled)

**Comment by mstrens** [comment]
So, U means that the ublox will be initialized by oXs (so normally u-center is not needed).
This was previously the only one option.
Note: if you select U option and your gps was already configured( and saved)  with u-center , oXs will sent his own parameters that could overwrite those that you had externally defined. Still the overwriting done by oXs are quite limited:
- rate = 10 hz (should perhaps be changed for best results)
- no nema msg
- 3 UBX msg (see above) enabled.

**Comment by Satcomix** [comment]
Hello Mstrens, 
That's a good  config feature with:
GPS=U
GPS=E
GPS=C
I will make some tests tomorow with my own U-Center config.txt for M8 and M10 with GPS, SBAS, Galileo und Glonass by 38400baud and 10Hz.
br,
Torsten

**Comment by fourfingers4** [comment]
Sounds great. 
That is best of two worlds, it will work both for users that don't want to get involved tinkering with u-center or have a gps without permanent config storage, and also users that want to setup their gps to their own taste.
Thanks a lot mstrens and Satcomix.
Can't wait to get back home and test it out.

**Comment by Satcomix** [comment]
Hello Mstrens,
As promised, I ran some tests.
TBS8-2 M8030
BN-220 M8030
BE-180 M10050
BE-220 M10050
BE-250 M10050
BE-280 M10050
SPG 2.01 to 5.10, PROTVER 15 to 34.10
Setting: GPS-U,
should only be used with the default setting for the M8030 chipset. Does not automatically switch to oXs_config ini. **Issue???**
The BE series automatically recognizes the ini and switches to oXs_config.
The new feature GPS=E, works with all tested GPS modules, with the setting 38400baud, 1 to 10Hz, an even faster NAVRate (18Hz) would work with the corresponding GNSS setting (1GNSS observe technical data from datasheet).
All GPS modules that can **save settings permanently** and have a working backup battery (BBR) can be operated with the GPS=E function.
br,
Torsten

**Comment by mstrens** [comment]
Thanks for the tests.
Sorry but I do not understand this sentence:
should only be used with the default setting for the M8030 chipset. Does not automatically switch to oXs_config ini. Issue???

Do you mean that U option does not work with e.g. a BN-220?

**Comment by Satcomix** [comment]
For example, if I set an M8030 module to NAVRate 5 Hz (to test GPS=E), this module does not automatically switch to 10Hz after the ini.
Usually default from M8 9600baud/1Hz.
Three UBX commands are required for the M10 module:
POSLLH, PVT, VELNED.
For the M8, in addition to these three, also SOL.

**Comment by Satcomix** [comment]
> Thanks for the tests. Sorry but I do not understand this sentence: should only be used with the default setting for the M8030 chipset. Does not automatically switch to oXs_config ini. Issue???
> 
> Do you mean that U option does not work with e.g. a BN-220?

When you use a default GPS BN180/220... BE180/220..... everything works with oXs_RP2040 GPS=U!!!

**Comment by mstrens** [comment]
This is not clear. You said:
For example, if I set an M8030 module to NAVRate 5 Hz (to test GPS=E), this module does not automatically switch to 10Hz by GPS=U after the ini.

In option GPS=U, oXs is supposed to send a command  
0xB5,0x62,0x06,0x08,0x06,0x00,0x64,0x00,0x01,0x00,0x01,0x00,0x7A,0x12, // NAV-RATE for 10 hz
So, I would expect that the M8 GPS would be set on 10 hz  even if it was configured before with NAVRate 5HZ

**Comment by Satcomix** [comment]
Make some more Test:
M10 setting: 38400baud/**5Hz** get the oXs ini and got 38400baud and **10Hz.**
This is also the result we expect!

**Comment by mstrens** [comment]
Not sure if it will help:  I put a version 2.0.1 where I send some commands twice (rate & types of ubx msg).

**Comment by Satcomix** [comment]
Some more tests:
After power up, no 10Hz flash, always 5Hz default, but after RESET the ini works and we get 10Hz flash.
That was the same problem with M10, when you set up to 2sec and  after that to 4ms back, would 10ms bring it to ini?

**Comment by mstrens** [comment]
Now version 2.0.2 with 2 sec instead of 4ms

**Comment by Satcomix** [comment]
No improvement in ver.2.0.2 after power off, available after reset 10Hz.

**Comment by Satcomix** [comment]
I've tried different NAVRATES.
5Hz,
3Hz,
2Hz,
everywhere the same problem.
After power down and on no ini with 10Hz.
After a reset come the 10Hz.
What else I noticed.
The blue LED normally goes on and off at 10Hz. But now it lights up like a flicker or at 20Hz?, but we've had it before.

**Comment by Satcomix** [comment]
It is a problem of the line in connection with the navrate.
If I set the navrate to 5Hz and the uart to 9600baud, everything is fine,
and the BN-220 synchronizes with 38400baud and 10Hz.
If I set the GPS to 38400baud and 5Hz, the initialization does not work.

**Comment by mstrens** [comment]
This could be an explanation.
oXs begins to send a command at 9600baud (because it expects that this is the default baudrate).
If the gps has been pre-configured at 38400 baud, it can't accept this command (to many errors) and probably disconnect his Rx pin for some delay.
I can try to increase the delay to 5 sec instead of 2 sec.
It is done in 2.0.3.

From my point of view, best is to say anyway that GPS baudrate should be the default value. If baudrate is set to another value (e.g. 115200), oXs will not be able to perform the initialisation

**Comment by Satcomix** [comment]
That is also my opinion,
if GPS=U is set when using an M8 e.g. BN-220, then a default(out of the box) GPS module must be connected.
With GPS=E, 38400 baud and xHz should be set.
It works with the BE (M10), regardless of whether GPS=U or GPS=E
br,
Torsten

**Comment by Satcomix** [comment]
I did some more tests and am very happy with version 2.0.3.
This version enables everyone with the **GPS=E** setting, to set their GPS module according to their own ideas (observe manufacturer information) in the GNSS and NAVRATE area.
A good piece of work, but it was worth it.
**Thank you for your work and effort.**
br,
Torsten

**Comment by mstrens** [comment]
here a version with 2 X 3sec delay instead of 2 X5 sec.
Perhaps it is enough to let M8 reactivate the RX signal

**Comment by Satcomix** [comment]
Hello Mstrens,
Unfortunately I'm busy with other things for today, so I won't be able to test the new version again until tomorrow morning.
br,
Torsten

**Comment by pierrotm777** [comment]
In GPS=U mode, my gps is found.

Protocol is SRXL2 (Spektrum)
CRSF baudrate = 420000
Voltage parameters:
    Scales : 1.000000 , 1.000000 , 1.000000 , 1.000000 
    Offsets: 0.000000 , 0.000000 , 0.000000 , 0.000000 
    One temperature sensor is connected on V3
RPM multiplier = 1.000000
Baro sensor is detected using MS5611
    Sensitivity min = 100 (at 100)   , max = 300 (at 1000)
    Hysteresis = 5 
Acc/Gyro is detected using MP6050
     Acceleration offsets X, Y, Z = 16694 , -718 , -16252
     Gyro offsets         X, Y, Z = -699 , 131 , 33
Airspeed sensor is not detected
First analog to digital sensor is detected using ads1115
    Measurement setup: 4 , 5 , 6 ,7
    Gains: 1 , 1 , 1 ,1
    Rates: 5 , 5 , 5 ,5
    Offsets: 0.000000 , 0.000000 , 0.000000 ,0.000000
    Scales: 1.000000 , 1.000000 , 1.000000 ,1.000000
    Averaged on: 10 , 10 , 10 ,10
Second analog to digital sensor is not detected
Foreseen GPS type is Ublox (configured by oXs) :GPS is detected but has not (yet) a fix
Failsafe type is HOLD

Config parameters are OK
Press ? + Enter to get help about the commands

 >>

**Comment by Satcomix** [comment]
Hello Pierre,
You must set the M8 Gps with U-Center to 38400baud and 5Hz. Then you install the new Version, and go to Terminal, and write GPS=U.  Power down the RP2040. 
If everything is ok, oxs will ini the modul with 38400baud and 10Hz.
Good luck.
greetings,
Torsten

**Comment by pierrotm777** [comment]
Baud 38400 and Navigation rate 10 [cyc]

Exact ?

**Comment by pierrotm777** [comment]
I have changed baudrate and navigation rate, shut down the gps and powered agin and it was always with 38400 and 10cyc.
The terminal return me:
Foreseen GPS type is Ublox (configured by oXs) :GPS is detected but has not (yet) a fix

**Comment by pierrotm777** [comment]
But if i return on u-center my gps is in 9600 bauds and 1 cyc.
May be my lipo backup is down ?

**Comment by Satcomix** [comment]
You must Set the M8 to 38400baud and 5Hz, and save the settings in GPS, or make a config.txt with U-Center and save it first on PC, and then playback to GPS, and save in BBR/Flash.
After Power down the oXs_on_RP2040 and Power on, the ini must work and set the GPS to 38400baud and 10Hz.

**Comment by pierrotm777** [comment]
My m8n don't save new baud value and stay always to 9600.

**Comment by Satcomix** [comment]
I also have some GPS, that do not save any setting.
Cheap quality from CN. :-(
Tomorow morning i can  make the tests.
I have to look if all versions from SPG3.01 to 5.10 and all Protversions to 34.10 will work.
Thank you.
br,
Torsten

**Comment by Satcomix** [comment]
Hello Mstrens,
I made the next tests with version 2.0.4.
The ini of an M8 38400baud/5Hz and M10 38400/1Hz works, but the data transmission doesn't seem to be completely clean to me.
If I connect a GPS with 9600/1Hz, I get a clean 10Hz flashing signal after the ini.
When initializing with a preset GPS M8/M10 38400/1Hz or 5Hz, 
i get a flickering almost like a permanent light after the ini, where I can't see the frequency.
br,
Torsten

**Comment by Satcomix** [comment]
The only thing that comes to mind for the GPS_TX would be an intermediate step, since everything works at 9600baud with 5Hz.
In the BE/BN series with default 38400baud/1Hz you would have to minimize the rate to 9600baud/1Hz in the first step and then synchronize it to 38400baud with 10Hz in order to get a clean signal.

**Comment by mstrens** [comment]
A nearly permanent light probably means that the GPS sent to many frames (probably that the command to cancel NEMA has not been handled.
I can go back to the version with 2 delays of 5 sec (instead the current version with 3 sec). You said that this version was ok.

**Comment by Satcomix** [comment]
I have the versions saved, which version should it be? 2.0.0??? or 2.0.1???

**Comment by Satcomix** [comment]
> A nearly permanent light probably means that the GPS sent to many frames (probably that the command to cancel NEMA has not been handled. I can go back to the version with 2 delays of 5 sec (instead the current version with 3 sec). You said that this version was ok.

I make a sniffing (GPS_TX) with U-Center, then i can see if NMEA is disabled.

**Comment by Satcomix** [comment]
You are right, NMEA is not disabled.

**Comment by mstrens** [comment]
I do not remember which version used 5 sec.
I made 2.0.5 with 5 sec.

I also reactivated in this version the possibility to use the boot button to set up the failsafe.
I disabled this function some weeks ago when I made other changes (using second core for sensors) because I was not sure it would work correctly. I hope it works (not tested my self).

**Comment by Satcomix** [comment]
NMEA is allready there!!!

**Comment by mstrens** [comment]
I think I understand.
I had to send twice the M8 command to disable NEMA (once at 9600 and once more at 38400 - for the case the GPS was already in 38400).
Now version 2.0.6
Still 2 X 5 sec but it could be that I can now reduce the second delay (to be tested afterwards)

**Comment by Satcomix** [comment]
2.0.6 works with 9600baud/1Hz , 38400baud/1Hz, 38400baud/5Hz M8 and M10 Modules

**Comment by Satcomix** [comment]
M8 works once with 9600/1Hz and 38400/5Hz then flashes 10Hz, everything OK,
but M10 initializes without NMEA but does not flash continuously with 10Hz anymore, as if NMEA were activated but it is not.
With version 2.0.5 everything was fine.

**Comment by mstrens** [comment]
now 2.0.8
back to 2X 5 sec (just like 2.0.8)

**Comment by Satcomix** [comment]
Sorry, I made a mistake and rechecked version 2.0.7.
It all works with M8 and M10 with 9600/1Hz, 38400/5Hz goto 38400baud with 10Hz blinking.

**Comment by mstrens** [comment]
Ok, no problem.
2.0.9 is back to 2.0.7 (5sec + 3 sec)

**Comment by Satcomix** [comment]
Everything works in Version 2.0.9(2.0.7) as it should, whether with M8 or M10 with different baud rates and NAV RATES.
br,
Torsten

**Comment by Satcomix** [comment]
Hello Mstrens,
I did some more tests with M8/BN and M10/BE GPS modules.
**Short question:**
The initialization for BZ Series with 115200baud/1Hz as you did in Ver.1.4.14 is no longer intended for the future, is it?
Everything else with M8/M10 and GPS=U/E works great!
br,
Torsten

**Comment by mstrens** [comment]
Currently, I do not plan to try to use 115200 baud.
Each time I send an init sequence at a different baud rate than expected by the GPS, it seems that it blocks the Rx pin and I have to wait 5 sec before sending other commands.
So, it is not a good solution to send commands at different baud rates.

**Comment by Satcomix** [comment]
Hello Mstrens,
Thanks for your answer.
I think that the current initialization is optimal for all users, thanks.
I'm about to install C-Make oXs without WIZ-IO, I still have some problems. I guess I'll have to read and ask more often.
Greetings,
Torsten

**Comment by Satcomix** [comment]
Hello Mstrens,
I have a problem creating a new build.
Issue:  GPS_IN_RECONFIGURATION;

switch (gpsState){
        case GPS_WAIT_END_OF_RESET:
            if (millisRp() > 1000) { //after x msec
                gpsState = GPS_IN_RECONFIGURATION;
                lastActionUs = 0;
                baudIdx = 0;       
            }
            break;
        case GPS_IN_RECONFIGURATION:
            if (lastActionUs == 0) {  // last action = 0 means that baudrate has to be rconfigure

Compiling .pio\build\raspberry-mstrens\src\gps.o
src\gps.cpp: In member function 'void GPS::handleGpsUblox()':
src\gps.cpp:293:28: error: 'GPS_IN_RECONFIGURATION' was not declared in this scope
                 gpsState = GPS_IN_RECONFIGURATION;
                            ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
*** [.pio\build\raspberry-mstrens\src\gps.o] Error 1

**Comment by pierrotm777** [comment]
Satcomix, delete your actual src directory and reload it !

**Comment by mstrens** [comment]
It looks like you do not have the latest version.

**Comment by Satcomix** [comment]
I don't know what happened when copying the 2.0.9 -main and -test.
Copied again to the Projects folder. Now everything is running under Platformio as usual, thanks to both of you.
